Kernel/VirtIO/VirtIOConsole.h

```diff
@@ -38,10 +38,39 @@ class VirtIOConsole final : public CharacterDevice
     virtual String device_name() const override { return String::formatted("hvc{}", minor()); }
     virtual void handle_queue_update(u16 queue_index) override;
 
-    OwnPtr<Region> m_receive_region;
-    OwnPtr<Region> m_transmit_region;
+    class VMRingBuf {
+    public:
+        VMRingBuf(const StringView region_name, size_t capacity);
+
+        bool has_space() const { return num_used_bytes < capacity_in_bytes; }
+        bool copy_data_in(const UserOrKernelBuffer& buffer, size_t length, FlatPtr& start_of_copied_data, size_t& bytes_copied);
+        void reclaim_space(FlatPtr region_start, size_t chunk_size)
+        {
+            VERIFY(start_of_used_paddr() == region_start);
+            VERIFY(num_used_bytes >= chunk_size);
+            num_used_bytes -= chunk_size;
+            start_of_used += chunk_size;
+            dbgln_if(VIRTIO_DEBUG, "Reclaimed {} bytes of space in console ringbuf", chunk_size);
+        }
+
+    private:
+        FlatPtr start_of_used_paddr() const
+        {
+            size_t start = start_of_used % capacity_in_bytes;
+            return m_region->physical_page(start / PAGE_SIZE)->paddr().offset(start % PAGE_SIZE).get();
+        }
+        OwnPtr<Region> m_region;
+        OwnPtr<WaitQueue> m_wait_queue;
+        size_t start_of_used {};
+        size_t num_used_bytes {};
+        size_t capacity_in_bytes;
+    };
+
+    OwnPtr<VMRingBuf> m_receive_buf;
+    OwnPtr<VMRingBuf> m_transmit_buf;
```

Comment:
theres no need to shorten this
```suggestion
    OwnPtr<VMRingBuf> m_receive_buffer;
    OwnPtr<VMRingBuf> m_transmit_buffer;
```